src/perception/aruco_processing.cpp

```diff
@@ -0,0 +1,136 @@
+        fidArray.header.frame_id = ODOM_FRAME;
+        fidArray.header.stamp = ros::Time::now();
+        fidArray.header.seq = mSeqNum;
+        fidArray.transforms.reserve(mPersistentFiducials.size());
+
+        // Add readings to the persistent representations of the fiducials
+        for (auto [id, immediateFid]: mImmediateFiducials) {
+            PersistentFiducial& fid = mPersistentFiducials[id];
+            if (!immediateFid.fidInCam.has_value()) continue; // This is set if the point cloud had a valid reading for this fiducial
+
+            fid.id = id;
+            SE3 fidInOdom = SE3::transform(mTfBuffer, ROVER_FRAME, ODOM_FRAME, immediateFid.fidInCam.value(), mSeqNum);
+            fid.setFilterParams(mFilterCount, mFilterProportion);
+            fid.addReading(fidInOdom);
+        }
+
+        // Send all transforms of persistent fiducials
+        for (auto [id, fid]: mPersistentFiducials) {
+            if (!fid.fidInOdomX.ready()) continue; // Wait until the filters have enough readings to become meaningful
+
+            SE3::sendTransform(mTfBroadcaster, "fiducial" + std::to_string(id), ODOM_FRAME, fid.getPose(), mSeqNum);
+        }
+
+        mFidPub.publish(fidArray);
+
+        size_t detectedCount = mIdsCache.size();
+        if (mIsVerbose || !mPrevDetectedCount.has_value() || detectedCount != mPrevDetectedCount.value()) {
+            mPrevDetectedCount = detectedCount;
+            ROS_INFO("Detected %zu markers", detectedCount);
+        }
+
+        if (!mImmediateFiducials.empty()) {
+            cv::aruco::drawDetectedMarkers(mCvPtr->image, mCornersCache, mIdsCache);
+        }
+
+        if (mPublishImages) {
+            mImgPub.publish(mCvPtr->toImageMsg());
+        }
+
+        mSeqNum++;
+    } catch (cv_bridge::Exception const& e) {
+        ROS_ERROR("cv_bridge exception: %s", e.what());
+    } catch (cv::Exception const& e) {
+        ROS_ERROR("cv exception: %s", e.what());
+    }
+}
+
+/**
+ * @brief       Retrieve the pose of the fiducial in camera space
+ * @param msg   3D Point Cloud with points stored relative to the camera
+ * @param u     X Pixel Position
+ * @param v     Y Pixel Position
+ */
+SE3 getFidInCamFromPixel(sensor_msgs::PointCloud2ConstPtr const& msg, size_t u, size_t v) {
+    // Could be done using PCL camToPoint clouds instead
+    size_t arrayPos = v * msg->row_step + u * msg->point_step;
+    size_t arrayPosY = arrayPos + msg->fields[0].offset;
+    size_t arrayPosZ = arrayPos + msg->fields[1].offset;
+    size_t arrayPosX = arrayPos + msg->fields[2].offset;
+
+    cv::Point3f point;
+    std::memcpy(&point.x, &msg->data[arrayPosX], sizeof(point.x));
+    std::memcpy(&point.y, &msg->data[arrayPosY], sizeof(point.y));
+    std::memcpy(&point.z, &msg->data[arrayPosZ], sizeof(point.z));
+
+    SE3 pointInCam{};
+    pointInCam.orientation.w = 1.0;
+    pointInCam.position.x = +point.x;
+    pointInCam.position.y = -point.y;
```

Comment:
please don't do shit like this without a comment 🙏😛

Reply:
Lol true. In fact this should be a ctor

Reply:
We actually need to figure this out cuz it seemed mad arbitrary, I only figured that out through trial and error looking at the values.